### PR TITLE
Update combobox and calendar with directives

### PR DESCRIPTION
**Pull request by reshaw**
Add directives for the ComboBox and Calendar to enable the Angular-style template localization pattern.

Change wrapper to fix some timing bugs

### CODE REVIEW

File: apps/docs/src/app/containers/component-docs/fabric/fabric.component.html [hunk 1]
```diff
@@ -22,7 +22,36 @@ <h2>Dialog</h2>
 >
   {{ sampleContent2 }} {{ sampleContent3 }}
   <fab-dialog-footer>
-    <fab-default-button (onClick)="clickSave()" text="Save" contentStyle="margin-right: 10px;"></fab-default-button>
```
Comment (bengry) [inline apps/docs/src/app/containers/component-docs/fabric/fabric.component.html]: Any reason this was removed?

File: libs/core/src/lib/components/wrapper-component.ts [hunk 2]
```diff
@@ -251,9 +253,9 @@ export abstract class ReactWrapperComponent<TProps extends {}> implements AfterV
     };
   }
 
-  private _passAttributesAsProps() {
+  private _passAttributesAsProps() { 
     const hostAttributes = Array.from((this.elementRef.nativeElement as HTMLElement).attributes);
-
+
```
Comment (bengry) [inline libs/core/src/lib/components/wrapper-component.ts]: Please make sure to re-format your code with Prettier when committing to avoid inconsistencies like this.

File: libs/core/src/lib/components/wrapper-component.ts [hunk 3]
```diff
@@ -97,7 +97,7 @@ export abstract class ReactWrapperComponent<TProps extends {}> implements AfterV
     this._contentClass = value;
     if (isReactNode(this.reactNodeRef.nativeElement)) {
       this.reactNodeRef.nativeElement.setProperty('className', classnames(value));
-      this.changeDetectorRef.detectChanges();
+      this.markForCheck();
```
Comment (bengry) [inline libs/core/src/lib/components/wrapper-component.ts]: Did you encounter any issues that warrant this change? `markForCheck()` does more than `this.changeDetectorRef.detectChanges()`.

File: libs/core/src/lib/components/wrapper-component.ts [hunk 4]
```diff
@@ -118,7 +118,7 @@ export abstract class ReactWrapperComponent<TProps extends {}> implements AfterV
     if (isReactNode(this.reactNodeRef.nativeElement)) {
       const stringValue = typeof value === 'string' ? value : stylenames(value);
       this.reactNodeRef.nativeElement.setProperty('style', toStyle(stringValue));
-      this.changeDetectorRef.detectChanges();
+      this.markForCheck();
```
Comment (bengry) [inline libs/core/src/lib/components/wrapper-component.ts]: Did you encounter any issues that warrant this change? `markForCheck()` does more than `this.changeDetectorRef.detectChanges()`.

File: libs/core/src/lib/components/wrapper-component.ts [hunk 5]
```diff
@@ -142,9 +142,11 @@ export abstract class ReactWrapperComponent<TProps extends {}> implements AfterV
     this._shouldSetHostDisplay = setHostDisplay;
   }
 
-  ngAfterViewInit() {
+  ngAfterContentInit() {
```
Comment (bengry) [inline libs/core/src/lib/components/wrapper-component.ts]: Why the change?
Also, if this stays - make sure to change the `implements` of this component respectively.

File: libs/fabric/src/lib/components/calendar/calendar.module.ts [hunk 6]
```diff
@@ -8,11 +8,12 @@ import * as CalendarCss from 'office-ui-fabric-react/lib-amd/components/Calendar
 import { Calendar } from 'office-ui-fabric-react';
 import { noop } from '../../utils/noop';
 import { FabCalendarComponent } from './calendar.component';
+import { CalendarStringsDirective } from './directives/calendar-strings-directive.component';
 
 // Dummy action to force CalendarCss to load and not be tree-shaken away.
 noop(CalendarCss);
 
-const components = [FabCalendarComponent];
+const components = [FabCalendarComponent, CalendarStringsDirective];
```
Comment (bengry) [inline libs/fabric/src/lib/components/calendar/calendar.module.ts]: Please rename to `declarations`, so it reflects the array's contents.

File: libs/fabric/src/lib/components/calendar/directives/calendar-strings-directive.component.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+// Licensed under the MIT License.
+
+import { Directive, Input } from '@angular/core';
+import { ICalendarStrings } from 'office-ui-fabric-react';
+
+/**
+ * Wrapper directive for calendar strings
+ */
+@Directive({ selector: 'fab-calendar > fab-calendar-strings' })
+export class CalendarStringsDirective {
```
Comment (bengry) [inline libs/fabric/src/lib/components/calendar/directives/calendar-strings-directive.component.ts]: Looks good.
Can you please add an example usage in the demo app?

File: libs/fabric/src/lib/components/combo-box/combo-box.module.ts [hunk 8]
```diff
@@ -7,8 +7,15 @@ import { NgModule, NO_ERRORS_SCHEMA } from '@angular/core';
 import { ComboBox, VirtualizedComboBox } from 'office-ui-fabric-react';
 import { FabComboBoxComponent } from './combo-box.component';
 import { FabVirtualizedComboBoxComponent } from './virtualized-combo-box.component';
+import { ComboBoxOptionDirective } from './directives/combo-box-option.directive';
+import { ComboBoxOptionsDirective } from './directives/combo-box-options.directive';
 
-const components = [FabComboBoxComponent, FabVirtualizedComboBoxComponent];
+const components = [
```
Comment (bengry) [inline libs/fabric/src/lib/components/combo-box/combo-box.module.ts]: Please rename to `declarations`

File: libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { Directive, Input } from '@angular/core';
+import { IComboBoxOption } from 'office-ui-fabric-react';
+
+/**
+ * Wrapper directive for creating a ComboBoxOption
+ */
+@Directive({ selector: 'fab-combo-option' })
```
Comment (bengry) [inline libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts]: How about renaming to `fab-combo-box-option` to stay consistant with the component's name?

File: apps/docs/src/app/containers/component-docs/fabric/fabric.component.html (hunk 1 above)
Comment (reshaw) [reply apps/docs/src/app/containers/component-docs/fabric/fabric.component.html]: Yes. I was troubleshooting the issue that resulted in the change in the wrapper from changeDetectorRef.detectChanges() to markForCheck(). I'll comment more there. Reverted this.

File: libs/core/src/lib/components/wrapper-component.ts (hunk 2 above)
Comment (reshaw) [reply libs/core/src/lib/components/wrapper-component.ts]: Acknowledged.

File: libs/core/src/lib/components/wrapper-component.ts (hunk 3 above)
Comment (reshaw) [reply libs/core/src/lib/components/wrapper-component.ts]: When first using the Calendar, I noticed that if it shared a template with another React-wrapped component, and that other component had any attributes that would later become property-bindings, that the Calendar would fail, having not correctly bound its own properties. I managed to work out that this was somehow related to the renderer. Changing this fixed the problem.

File: libs/core/src/lib/components/wrapper-component.ts (hunk 4 above)
Comment (reshaw) [reply libs/core/src/lib/components/wrapper-component.ts]: See other reply on similar change

File: libs/core/src/lib/components/wrapper-component.ts (hunk 5 above)
Comment (reshaw) [reply libs/core/src/lib/components/wrapper-component.ts]: I am actually hoping for your insight on this. The following conversation applies to the Calendar and ComboBox.

The component was not binding properties from the directive properly when the passAttributesAsProps call was in ngAfterContentInit, but it would only do this in the specific case where the binding was only coming from the directive, and not from an attribute. If we also set the property directly through attribute-binding, then Angular binds the React properties correctly. This is not ideal, sort of messy, and not how anyone would reasonably use the component.

After some investigation, Jerry and I came to the conclusion that it may be a timing issue. If the passAttributesAsProps function call happened early and late (in ngOnChanges, which would happen with attribute-binding), then everything is fine. If the call just happened late (in ngAfterContentInit), then the binding would not happen correctly, and the event callbacks would fire with incorrect code. Moving the passAttributesAsProps function call to happen earlier in the Angular lifecycle fixed the problem.

File: libs/fabric/src/lib/components/calendar/calendar.module.ts (hunk 6 above)
Comment (reshaw) [reply libs/fabric/src/lib/components/calendar/calendar.module.ts]: Done.

File: libs/fabric/src/lib/components/combo-box/combo-box.module.ts (hunk 8 above)
Comment (reshaw) [reply libs/fabric/src/lib/components/combo-box/combo-box.module.ts]: Done.

File: libs/fabric/src/lib/components/calendar/directives/calendar-strings-directive.component.ts (hunk 7 above)
Comment (reshaw) [reply libs/fabric/src/lib/components/calendar/directives/calendar-strings-directive.component.ts]: Example added to demo.

File: libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts (hunk 9 above)
Comment (reshaw) [reply libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts]: Done.

File: apps/demo/src/app/app.component.html [hunk 10]
```diff
@@ -142,6 +142,31 @@ <h2>Getting up and running...</h2>
 
   <fab-calendar [strings]="strings" (onSelectDate)="onSelectDate($event)"></fab-calendar>
 
+  <fab-calendar (onSelectDate)="onSelectDate($event)">
+    <fab-calendar-strings
+      key="key"
```
Comment (bengry) [inline apps/demo/src/app/app.component.html]: why is the key needed?

File: apps/docs/src/app/containers/component-docs/fabric/fabric.component.html [hunk 11]
```diff
@@ -22,7 +22,47 @@ <h2>Dialog</h2>
...
+  </options>
+</fab-combo-box>
+{{ selectedComboBoxKey }}: {{ selectedComboBoxValue }}
+
+<hr />
+
+<h2>Directive Calendar</h2>
+<fab-calendar (onSelectDate)="onSelectDate($event)">
+  <fab-calendar-strings
+    key="key"
```
Comment (bengry) [inline apps/docs/src/app/containers/component-docs/fabric/fabric.component.html]: Same as the other comment - why is `key` needed?

File: libs/core/src/lib/components/wrapper-component.ts (hunk 5 above)
Comment (bengry) [reply libs/core/src/lib/components/wrapper-component.ts]: I'm not sure I understand why passing it later than you can (i.e. in anything later than `ngOnChanges`) needed is an issue.

File: libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+import { Directive, Input } from '@angular/core';
+import { IComboBoxOption } from 'office-ui-fabric-react';
+
+/**
+ * Wrapper directive for creating a ComboBoxOption
+ */
+@Directive({ selector: 'fab-combo-box-option' })
+export class ComboBoxOptionDirective {
+  @Input() optionKey: IComboBoxOption['key'];
```
Comment (bengry) [inline libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts]: Why not call it `key` in the property name too? We'd like to stay consistant with the `office-ui-fabric-react` names as much as possible, especially with stuff like #91   in the pipeline.

File: libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+  @Input() text: IComboBoxOption['text'];
+  @Input() title?: IComboBoxOption['title'];
+  @Input() itemType: IComboBoxOption['itemType'];
+  @Input() index?: IComboBoxOption['index'];
+  @Input() ariaLabel?: IComboBoxOption['ariaLabel'];
+  @Input() selected?: IComboBoxOption['selected'];
+  @Input() disabled?: IComboBoxOption['disabled'] = false;
```
Comment (bengry) [inline libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts]: why the need for a default value? Let `office-ui-fabric-react` handle it.

File: libs/core/src/lib/components/wrapper-component.ts (hunk 5 above)
Comment (reshaw) [reply libs/core/src/lib/components/wrapper-component.ts]: ngOnChanges doesn't fire consistently.

File: libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts (hunk 12 above)
Comment (reshaw) [reply libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts]: I concur, but that runs foul of the wrapper-component _isForbiddenAttribute method. This causes the ComboBox to misbehave (it ignores the key attribute binding and triggers a console warning).

File: apps/demo/src/app/app.component.html (hunk 10 above)
Comment (reshaw) [reply apps/demo/src/app/app.component.html]: It's not, removed.

File: apps/docs/src/app/containers/component-docs/fabric/fabric.component.html (hunk 11 above)
Comment (reshaw) [reply apps/docs/src/app/containers/component-docs/fabric/fabric.component.html]: It's not, removed.

File: libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts (hunk 13 above)
Comment (reshaw) [reply libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts]: Concur.

File: libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts (hunk 12 above)
Comment (bengry) [reply libs/fabric/src/lib/components/combo-box/directives/combo-box-option.directive.ts]: Makes sense, let's keep it like that then. Please do add some JDoc on this to make a note of this.

File: libs/core/src/lib/components/wrapper-component.ts (hunk 5 above)
Comment (bengry) [reply libs/core/src/lib/components/wrapper-component.ts]: Right, but what about the difference between `ngAfterViewInit` vs `ngAfterContentInit` in this case?

Comment (reshaw) [reply libs/core/src/lib/components/wrapper-component.ts]: Firing passAttributesAsProps at ngAfterViewInit time works. Firing it at ngAfterContentInit causes event bindings to bind incorrectly.
